File: 376.摆动序列.cpp
// 官方题解
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        int n = nums.size();
        int down = 1, up = 1;
        for(int i = 1; i < n; ++i){
            if(nums[i] - nums[i - 1] > 0){
                up = down + 1;
            } else if (nums[i] - nums[i - 1] < 0){
                down = up + 1;
            }
        }
        return max(up, down);
    }
};


class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        int up = wiggle(nums, true);
        int down = wiggle(nums, false);
        return max(up, down);
    }

    int wiggle(vector<int>& nums, bool isUp){
        int len = 0;
        int n = nums.size();
        if(n == 0)
            return 0;
        for(int i = 1; i < n; ++i){
            if(isUp && nums[i] - nums[i - 1] > 0){
                ++len;
                isUp = !isUp;
            } else if(!isUp && nums[i] - nums[i - 1] < 0) {
                ++len;
                isUp = !isUp;
            }
        }
        return len + 1;
    }
};

File: 376.摆动序列.java
// 自己写的dp
class Solution {
    public int wiggleMaxLength(int[] nums) {
        if(nums.length < 2)
            return nums.length;
        int n = nums.length;
        int[] positive = new int[n];
        int[] negative = new int[n];
        for(int i = 0; i < n; ++i){
            positive[i] = 1;
            negative[i] = 1;
        }
        int maxp = 1, maxn = 1;
        for(int i = 1; i < n; ++i){
            if(nums[i] - nums[i - 1] > 0){
                positive[i] = maxn + 1;
                maxp = Math.max(maxp, positive[i]);
            } else if(nums[i] - nums[i - 1] < 0){
                negative[i] = maxp + 1;
                maxn = Math.max(maxn, negative[i]);
            }
        }
        return Math.max(maxp, maxn);
    }
}

// 评论区大神的解
class Solution {
    public int wiggleMaxLength(int[] nums) {
        int n = nums.length;
        if (n < 2) {
            return n;
        }
        int up = 1;
        int down = 1;
        for (int i = 1; i < n; i++) {
            if (nums[i] > nums[i - 1]) {
                up = down + 1;
            }
            if (nums[i] < nums[i - 1]) {
                down = up + 1;
            }
        }
        return Math.max(up, down);
    }
}